src/tests/iostreamtests.cc

```diff
@@ -0,0 +1,202 @@
+
+    char* ptr = buf->get_ptr(0);
+    unsigned len = buf->get_contiguous_length(ptr);
+    assert(memcmp(ptr, msg, len) == 0);
+
+    assert(stream.flush_nx() == sizeof(msg) - 1);
+    assert(buf->get_length() == 0);
+
+    std::vector<char> wdata;
+    bp_sys::extract_written_data(fd, wdata);
+
+    assert(wdata.size() == sizeof(msg) - 1);
+    assert(memcmp(wdata.data(), msg, wdata.size()) == 0);
+}
+
+void ostream_types_test()
+{
+    int fd = bp_sys::allocfd();
+    dio::ostream stream(fd);
+    dio::streambuf *buf = stream.get_buf();
+    assert(buf != nullptr);
+    assert(stream.good());
+
+    const char msg1[] = "The 2 + 2 equals to: ";
+    const int msg2 = 2 + 2;
+    const char msg_full[] = "The 2 + 2 equals to: 4";
+
+    assert(stream.write(msg1, msg2));
+    assert(buf->get_length() == sizeof(msg_full) - 1);
+
+    char* ptr = buf->get_ptr(0);
+    unsigned len = buf->get_contiguous_length(ptr);
+    assert(memcmp(ptr, msg_full, len) == 0);
+
+    assert(stream.flush_nx() == sizeof(msg_full) - 1);
+    assert(buf->get_length() == 0);
+
+    std::vector<char> wdata;
+    bp_sys::extract_written_data(fd, wdata);
+
+    assert(wdata.size() == sizeof(msg_full) - 1);
+    assert(memcmp(wdata.data(), msg_full, wdata.size()) == 0);
+}
+
+void ostream_large_msg_test()
+{
+    int fd = bp_sys::allocfd();
+    dio::ostream stream(fd);
+    dio::streambuf *buf = stream.get_buf();
+    assert(buf != nullptr);
+    assert(stream.good());
+
+    char msg[IOSTREAM_BUFSIZE + 2];
+    std::fill_n(msg, IOSTREAM_BUFSIZE + 1, 'a');
+    msg[IOSTREAM_BUFSIZE + 1] = '\0';
+
+    assert(stream.write(msg));
+    assert(buf->get_length() == 0);
+
+    std::vector<char> wdata;
+    bp_sys::extract_written_data(fd, wdata);
+
+    assert(wdata.size() == sizeof(msg) - 1);
+    assert(memcmp(wdata.data(), msg, wdata.size()) == 0);
+}
+
+void istream_basic_test()
+{
+    int fd = bp_sys::allocfd();
+    dio::istream stream(fd);
+    dio::streambuf *buf = stream.get_buf();
+    assert(buf != nullptr);
+    assert(stream.good());
+
+    std::vector<char> wdata = { 'L', '1', '\n', 'L','2', '\n', '\n', 'L', '3' };
+    bp_sys::supply_read_data(fd, wdata);
+    std::string line;
+
+    assert(dio::get_line(stream, line) == 2);
+    assert(line.compare("L1") == 0);
+
+    assert(dio::get_line(stream, line) == 2);
+    assert(line.compare("L2") == 0);
+
+    assert(dio::get_line(stream, line) == 1);
+    assert(line.size() == 0);
+
+    assert(dio::get_line(stream, line) == 2);
+    assert(line.compare("L3") == 0);
+
+    assert(dio::get_line_until_eof(stream, line) == 0);
+    assert(stream.eof());
+    assert(line.compare("L3") == 0);
+}
+
+void istream_file_open_test()
+{
+    std::vector<char> wdata = { 'L', '1', '\n', 'L','2', '\n', '\n', 'L', '3' };
+    bp_sys::supply_file_content("file", std::move(wdata));
+
+    dio::istream stream;
+    assert(stream.open_nx("file"));
+    assert(stream.is_open());
+    assert(stream.get_fd() >= 0);
```

Comment:
This test is basically the same as the previous one, I think they should just be combined.

It would be good to have a separate test of a `getline` read which reads across the buffer boundary, that would have found the bug in `get_line_nx()`.

Reply:
Take a look at new `istream_buffer_boundary_test` test.